### PR TITLE
Replace primitive Array fcalls with managed implementations

### PR DESCRIPTION
**Pull request by jkotas**
Prep-work for #27106

### CODE REVIEW

File: src/vm/reflectioninvocation.cpp [hunk 1]
```diff
@@ -2082,159 +2082,6 @@ FCIMPL0(FC_BOOL_RET, ReflectionInvocation::TryEnsureSufficientExecutionStack)
...
-
-LONG SODetectionFilter(EXCEPTION_POINTERS *ep, void* pv)
-{
-    WRAPPER_NO_CONTRACT;
-    DefaultCatchFilterParam param(COMPLUS_EXCEPTION_EXECUTE_HANDLER);
-    if (DefaultCatchFilter(ep, &param) == EXCEPTION_CONTINUE_EXECUTION)
-    {
-        return EXCEPTION_CONTINUE_EXECUTION;
-    }
-
```
Comment (jkotas) [inline src/vm/reflectioninvocation.cpp]: I have fixed and cleanup up a few other things as I have worked on the change.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 2]
```diff
@@ -170,7 +162,15 @@ internal static bool IsBitwiseEquatable<T>()
 
         [Intrinsic]
         internal static ref byte GetRawSzArrayData(this Array array) =>
-            ref Unsafe.As<RawSzArrayData>(array).Data;
+            ref Unsafe.As<RawArrayData>(array).Data;
+
+        internal static unsafe ushort GetElementSize(this Array array)
+        {
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: The inlineable managed impl will be needed to get a good performance out of managed BlockCopy impl.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 3]
```diff
@@ -65,6 +65,30 @@ public static object GetUninitializedObject(Type type)
...
+            }
+            catch
+            {
+                exceptionThrown = true;
+                throw;
+            }
+            finally
+            {
+                backoutCode(userData, exceptionThrown);
+            }
```
Comment (VSadov) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs]: NIT: perhaps can initialize exceptionThrown to true and then set to false right before leaving try. No need for catch/rethrow then. Not sure if that is any better for execution, but it wold be shorter somewhat.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs [hunk 4]
```diff
@@ -94,9 +94,12 @@ internal class RawData
...
-    internal class RawSzArrayData
+    internal class RawArrayData
     {
-        public IntPtr Count; // Array._numComponents padded to IntPtr
+        public uint Length; // Array._numComponents padded to IntPtr
+#if BIT64
+        public uint Padding;
+#endif
         public byte Data;
     }
```
Comment (EgorBo) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs]: Maybe worth moving it to the shared? Along with new Length and LongLength.

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs]: Maybe eventually. There is a lot of runtime specific quirks.

Does Mono use the same memory layout for array as CoreCLR?

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 2 above)
Comment (VSadov) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Yes, `GetObjectMethodTablePointer` is a bit of a problem.
This is just reinterpreting `object` as `IntPtr`, but whatever combination of `Unsafe` or unsafe math I use, I end up with JIT introducing a temp.
Perhaps another ` ldarg.0; conv.u; ret` intrinsic is needed.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 5]
```diff
@@ -65,6 +65,26 @@ public static object GetUninitializedObject(Type type)
             return GetUninitializedObjectInternal(type);
         }
 
+        public static void ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, object userData)
```
Comment (marek-safar) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs]: nit: TryCode and CleanupCode delegates uses `object?` so it should match to userData

Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs]:
```suggestion
        public static void ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, object? userData)
```

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Good catch. The old signature had a ?... looks like it just got lost in the move.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 2 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Yes, there is one extra instruction, but it is not a big deal.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs (hunk 4 above)
Comment (filipnavara) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs]: Not the same layout but similar one. Length/Data is exposed to the managed code in the same way.

**Comment by jkotas** [comment]
@dotnet/dnceng Many CoreCLR legs are failing Azure DevOps service unavailable errors. Could you please take a look?

```
error : Request to https://dev.azure.com/dnceng/public/_apis/test/runs?api-version=5.0-preview.2 returned failed status 503 Service Unavailable [/__w/3/s/tests/helixpublishwitharcade.proj]
error :  [/__w/3/s/tests/helixpublishwitharcade.proj]
error : {"$id":"1","innerException":null,"message":"TF10216: Azure DevOps services are currently unavailable. Try again later. Activity Id: 4f1c656e-fbe6-47e4-8eaa-4d5b701cd8dc","typeName":"Microsoft.VisualStudio.Services.CircuitBreaker.CircuitBreakerShortCircuitException, Microsoft.VisualStudio.Services.Common","typeKey":"CircuitBreakerShortCircuitException","errorCode":0,"eventId":3000} [/__w/3/s/tests/helixpublishwitharcade.proj]
error : (NETCORE_ENGINEERING_TELEMETRY=Test) Request to https://dev.azure.com/dnceng/public/_apis/test/runs?api-version=5.0-preview.2 returned failed status 503 Service Unavailable
```

**Comment by stephentoub** [comment]
corefx CI is also struggling:
https://github.com/dotnet/corefx/pull/41749

**Pull request merged**